{% load humanize %}
<head>
  <!-- this will only work if the CSS is defined within the template -->
  <style>
    body {
      font-size: 12px;
    }
    table {
      word-wrap: break-word;
      width: 100%;
      border-spacing: 0;
      font-size: 12px;
    }
    thead {
      display: table-row-group;
    }
    th, td {
      vertical-align: top;
      text-align: left;
      border-bottom: 1px solid #000;
    }
    th {
      border-top: 1px solid #000;
      background-color: #D3D3D3;
    }
    tr {
      page-break-before: always;
      page-break-after: always;
      page-break-inside: avoid;
    }
    .alert {
      padding: 10px;
      background-color: #f44336; /* Red */
      color: white;
      margin-bottom: 5px;
      text-align: center; 
    }
  </style>
</head>

<body>

  <h1>{{ sample_data.panel_obj.get_assay_display }} report for {{ sample_data.sample_id }}</h1>

  <!-- Fail warning box -->
  {% if sample_data.status == 'Analysis fail' %}
  <div class="alert">
    <h2><b>This sample has failed analysis</b></h2>
  </div> 
  <br>
  {% endif %}

  {% if sample_data.status == 'Bioinformatics QC fail' %}
  <div class="alert">
    <h2><b>This sample has failed bioinformatics QC</b></h2>
  </div> 
  <br>
  {% endif %}

  <!-- patient summary box -->
  <h2>Patient details</h2>
  <table>
    <thead>
      <tr>
        <th style="width: 20%">Lab number</th>
        <th style="width: 15%">Worksheet</th>
        <th style="width: 15%">Panel</th>
        <th style="width: 30%">Run ID</th>
        <th style="width: 20%">Genome Build</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>{{ sample_data.sample_id }}</td>
        <td>{{ sample_data.worksheet_id }}</td>
        <td>{{ sample_data.panel_obj }}</td>
        <td>{{ sample_data.run_id }}</td>
        <td>GRCh{{ sample_data.genome_build }}</td>
      </tr>
    </tbody>
  </table>
  <br>
  <br>

  {% if sample_data.status != 'Bioinformatics QC fail' %}

  <!-- SNV section - variant calls ------------------------------>
  {% if sample_data.panel_obj.show_snvs %}
  <h2>SNVs & indels - variant calls</h2>
  <small class="form-text text-muted">Only genuine variants are displayed here, any miscalled or failed calls are listed at the end of this report</small>

  <table>
    <thead>
      <tr>
        <th style="width: 20%">Variant</th>
        <th style="width: 10%">Gene</th>
        <th style="width: 25%">HGVS</th>
        <th style="width: 5%">VAF</th>
        <th style="width: 10%">IGV</th>
        <th style="width: 30%">Comments</th>
      </tr>
    </thead>
    <tbody>
    <!-- say no calls if empty -->
    {% if variant_data.no_calls %}
      <tr><td colspan="7">No calls</td></tr>

    <!-- otherwise list all genuine variants -->
    {% else%}
      {% for v in variant_data.variant_calls %}
        {% if v.final_decision == "Genuine" %}
        <tr>
          <td>{{ v.genomic | truncatechars:20 }}
          {% if v.manual_upload %}
        	Manual
          {% endif %}
          </td>
          <td>{{ v.gene }} ({{ v.exon }})</td>
          <td>{{ v.hgvs_c }} ({{ v.hgvs_p }})</td>
          {% if sample_data.panel_obj.assay == '3' %} <!-- 2dp for ctDNA, whole number for the rest-->
          <td>{{ v.vaf.vaf }}%</td>
          {% else %}
          <td>{{ v.vaf.vaf_rounded }}%</td>
          {% endif %}
          <td>{{ v.final_decision }}</td>
          <td>
            {% for c in v.comments %}
            <i>{{ c.user }}:</i><br>{{ c.comment | wordwrap:50 | linebreaksbr }}<br>
            {% endfor %}
          </td>
        </tr>
        {% endif %}
      {% endfor %}

    {% endif %}
    </tbody>
  </table>
  <br>
  <br>

  <!-- SNV section - coverage ------------------------------>
  <h2>Gene coverage</h2>

  <!-- gene overall coverage table - all panels -->
  <table>
    <thead>
      <tr>
        <th>Gene</th>
        {% if '135' in coverage_data.depth_cutoffs %}
        <th>Percent 135X</th>
        {% endif %}
        {% if '270' in coverage_data.depth_cutoffs %}
        <th>Percent 270X</th>
        {% endif %}
        {% if '500' in coverage_data.depth_cutoffs %}
        <th>Percent 500X</th>
        {% endif %}
        {% if '1000' in coverage_data.depth_cutoffs %}
        <th>Percent 1000X</th>
        {% endif %}
      </tr>
    </thead>

    <tbody>
      {% for gene, values in coverage_data.regions.items %}
      <tr>
        <td>{{ gene }}</td>
        {% if '135' in coverage_data.depth_cutoffs %}
        <td>{{ values.percent_135x }}%</td>
        {% endif %}
        {% if '270' in coverage_data.depth_cutoffs %}
        <td>{{ values.percent_270x }}%</td>
        {% endif %}
        {% if '500' in coverage_data.depth_cutoffs %}
        <td>{{ values.percent_500x }}%</td>
        {% endif %}
        {% if '1000' in coverage_data.depth_cutoffs %}
        <td>{{ values.percent_1000x }}%</td>
        {% endif %}
      </tr>
      {% endfor %}
    </tbody>
  </table>
  <br>
  <br>

  <!-- gaps summary for myeloid referrals-->
  {% if myeloid_coverage_summary %}
  <h2>Gaps summary</h2>

  <table>
    <thead>
      <tr>
        <th>Regions with...</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td style="width: 25%">0% coverage</td>
        <td style="width: 75%">{{ myeloid_coverage_summary.summary_0x }}</td>
      </tr>
      <tr>
        <td>&lt;270X depth</td>
        <td>{{ myeloid_coverage_summary.summary_270x }}</td>
      </tr>
    </tbody>
  </table>

  <!-- hotspot gaps for non-myeloid referrals -->
  {% else %}
  <h2>Hotspot gaps</h2>

  {% if '135' in coverage_data.depth_cutoffs %}
  <table>
    <thead>
      <tr>
        <th style="width: 50%">Hotspot gaps at 135X</th>
        <th style="width: 25%">Percent COSMIC</th>
        <th style="width: 25%">Counts COSMIC</th>
      </tr>
    </thead>

    <tbody>
    {% if coverage_data.gaps_present_135 %}
      {% for gene, values in coverage_data.regions.items %}
        {% for gap in values.gaps_135 %}
        <tr>
          <td>{{ gap.gene }}({{ gap.hgvs_transcript }}):{{ gap.hgvs_c }}</td>
          <td>{{ gap.percent_cosmic }}</td>
          <td>{{ gap.counts_cosmic }}</td>
        </tr>
        {% endfor %}
      {% endfor %}
    {% else %}
      <tr><td colspan="3">No hotspot gaps</td></tr>
    {% endif %}
    </tbody>
  </table>
  <br>
  {% endif %}

  {% if '270' in coverage_data.depth_cutoffs %}
  <table>
    <thead>
      <tr>
        <th style="width: 50%">Hotspot gap at 270X</th>
        <th style="width: 25%">Percent COSMIC</th>
        <th style="width: 25%">Counts COSMIC</th>
      </tr>
    </thead>

    <tbody>
    {% if coverage_data.gaps_present_270 %}
      {% for gene, values in coverage_data.regions.items %}
        {% for gap in values.gaps_270 %}
        <tr>
          <td>{{ gap.gene }}({{ gap.hgvs_transcript }}):{{ gap.hgvs_c }}</td>
          <td>{{ gap.percent_cosmic }}</td>
          <td>{{ gap.counts_cosmic }}</td>
        </tr>
        {% endfor %}
      {% endfor %}
    {% else %}
      <tr><td colspan="3">No hotspot gaps</td></tr>
    {% endif %}
    </tbody>
  </table>
  <br>
  {% endif %}

  {% if '500' in coverage_data.depth_cutoffs %}
  <table>
    <thead>
      <tr>
        <th style="width: 50%">Hotspot gaps at 500X</th>
        <th style="width: 25%">Percent COSMIC</th>
        <th style="width: 25%">Counts COSMIC</th>
      </tr>
    </thead>

    <tbody>
    {% if coverage_data.gaps_present_500 %}
      {% for gene, values in coverage_data.regions.items %}
        {% for gap in values.gaps_500 %}
        <tr>
          <td>{{ gap.gene }}({{ gap.hgvs_transcript }}):{{ gap.hgvs_c }}</td>
          <td>{{ gap.percent_cosmic }}</td>
          <td>{{ gap.counts_cosmic }}</td>
        </tr>
        {% endfor %}
      {% endfor %}
    {% else %}
      <tr><td colspan="3">No hotspot gaps</td></tr>
    {% endif %}
    </tbody>
  </table>
  <br>
  {% endif %}

  {% if '1000' in coverage_data.depth_cutoffs %}
  <table>
    <thead>
      <tr>
        <th style="width: 50%">Hotspot gaps at 1000X</th>
        <th style="width: 25%">Percent COSMIC</th>
        <th style="width: 25%">Counts COSMIC</th>
      </tr>
    </thead>

    <tbody>
    {% if coverage_data.gaps_present_1000 %}
      {% for gene, values in coverage_data.regions.items %}
        {% for gap in values.gaps_1000 %}
        <tr>
          <td>{{ gap.gene }}({{ gap.hgvs_transcript }}):{{ gap.hgvs_c }}</td>
          <td>{{ gap.percent_cosmic }}</td>
          <td>{{ gap.counts_cosmic }}</td>
        </tr>
        {% endfor %}
      {% endfor %}
    {% else %}
      <tr><td colspan="3">No hotspot gaps</td></tr>
    {% endif %}
    </tbody>
  </table>
  <br>
  {% endif %}

  {% endif %}

  <br>
  <br>
  {% endif %}

  {% if sample_data.panel_obj.show_fusions %}
  <h2>Fusions</h2>
  <small class="form-text text-muted">Only genuine fusions are displayed here, any miscalled or failed calls are listed at the end of this report</small>

  <table>
    <thead>
      <tr>
        <th style="width: 12%">Fusion</th>
        <th style="width: 7%">Fusion supporting reads</th>
        {% if sample_data.panel_obj.show_fusion_vaf %}<th style="width: 7%">VAF</th>{% endif %}
        <th style="width: 12%">Breakpoints</th>
        <th style="width: 25%">HGVS</th>
        <th style="width: 8%">IGV</th>
        <th style="width: 28%">Comments</th>
      </tr>
    </thead>

    <tbody>
    {% if fusion_data.no_calls %}
      {% if sample_data.panel_obj.show_fusion_vaf %}<tr><td colspan="7">No calls</td></tr>
      {% else %}<tr><td colspan="6">No calls</td></tr>
      {% endif %}

    {% else %}
      {% for v in fusion_data.fusion_calls %}
      {% if v.final_decision == "Genuine" %}
      <tr>
        <td>{{ v.fusion_genes }}</td>
        <td>{{ v.fusion_supporting_reads }}</td>
        {% if sample_data.panel_obj.show_fusion_vaf %}<td>{{ v.vaf }}%</td>{% endif %}
        <td>{{ v.left_breakpoint }}<br>{{ v.right_breakpoint }}</td>
        <td>
          <!-- loop required because bug in xhtml will collpase whole column if hgvs is None -->
          {% if v.fusion_hgvs %}
          {{ v.fusion_hgvs }}
          {% else %}
          N/A
          {% endif %}
        </td>
        <td>{{ v.final_decision }}</td>
        <td>
          {% for c in v.comments %}
          <i>{{ c.user }}: </i>{{ c.comment | wordwrap:50 | linebreaksbr }}<br>
          {% endfor %}
        </td>
      </tr>
      {% endif %}
      {% endfor %}
    {% endif %}
    </tbody>
  </table>
  <br>
  <br>

  <!-- fusion coverage -->
    {% if sample_data.panel_obj.show_fusion_coverage %}
    <h2>Coverage</h2>
    <table>
      <thead>
        <tr>
          <th>Number of reads in sample</th>
          <th>Number of reads in NTC</th>
          <th>Percent NTC</th>
        </tr>
      </thead>
  
      <tbody>
        <tr>
          <td>{{ sample_data.total_reads | intcomma }}</td>
          <td>{{ sample_data.total_reads_ntc | intcomma }}</td>
          <td>{{ sample_data.percent_reads_ntc }}%</td>
        </tr>
      </tbody>
    </table>
    <br>
    <br>
    {% endif %}

  {% endif %}

  {% endif %}

  <!-- checks section ------------------------------>
  <h2>Checks performed</h2>

  <table>
    <thead>
      <tr>
        <th style="width: 10%">Check</th>
        <th style="width: 20%">User</th>
        <th style="width: 10%">Pass/fail</th>
        {% if sample_data.status != 'Bioinformatics QC fail' %}<th style="width: 15%">Checks</th>{% endif %}
        <th style="width: 45%">Comments</th>
      </tr>
    </thead>

    <tbody>
      {% for c in sample_data.checks.all_checks %}
      <tr>
        <td>{{ forloop.counter }}</td>
        <td>{{ c.user }}<br>{{ c.signoff_time }}</td>
        <td>{% if c.get_status_display == "Complete" %}Pass{% else %}{{ c.get_status_display }}{% endif %}</td>
        {% if sample_data.status != 'Bioinformatics QC fail' %}
        <td>
          <i>Patient info check:</i> {{ c.patient_info_check }}
        {% if sample_data.panel_obj.show_snvs %}
          <br><i>NTC check:</i> {{ c.coverage_ntc_check }}
        {% endif %}
        {% if sample_data.panel_obj.manual_review_required %}
          <br><i>Manual variant check:</i> {{ c.manual_review_check }}
        {% endif %}
        </td>
        {% endif %}
        <td>
            {% if sample_data.status != 'Bioinformatics QC fail' %}<i>Coverage:</i> {{ c.coverage_comment | wordwrap:50 | linebreaksbr }}<br>{% endif %}
          <i>Overall:</i> {{ c.overall_comment | wordwrap:50 | linebreaksbr }}</td>
      </tr>
      {% endfor %}
    </tbody>

  </table>
  <br>
  
  <!-- footer -->
  <!-- list any miscalled or failed variants -->
  {% if sample_data.panel_obj.show_snvs and sample_data.status != 'Bioinformatics QC fail'%}
  <p class="form-text text-muted">Miscalled or failed SNV/indels, please refer back to the SVD to see more detail: {{ variant_data.other_calls_text }}</p>
  {% endif %}

  <!-- list any miscalled or failed fusions -->
  {% if sample_data.panel_obj.show_fusions %}
  <p class="form-text text-muted">Miscalled or failed fusions, please refer back to the SVD to see more detail: {{ fusion_data.other_calls_text }}</p>
  {% endif %}

  <!-- audit trail -->
  <p>View report in somatic variant database: <a href="http://10.69.115.27/svd/analysis/{{ sample_data.sample_pk }}#report" target="_blank">http://10.69.115.27/svd/analysis/{{ sample_data.sample_pk }}#report</a></p>
  <p>PDF report generated {% now "jS F Y H:i" %}</p>

</body>
